### PR TITLE
Added ability to save plots

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #6183
I'll add tests after we merge Notebook Serializer (much much simpler than creating *.ipynb files, opening, etc)

Corresponding change https://github.com/microsoft/vscode-notebook-renderers/pull/55

saveAs

### CODE REVIEW

File: src/client/datascience/notebook/outputs/rendererCommunication.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public activate() {
+        if (!this.useVSCNotebook) {
+            return;
+        }
+
+        const api = notebooks.createRendererMessaging<unknown, RendererMessageTypes>(rendererId);
+        api.onDidReceiveMessage(
+            ({ editor, message }) => {
+                if (message.type === 'saveAs') {
+                    this.plotSaveHandler.savePlot(editor, message.outputId, message.mimnType).catch(noop);
```
Comment (joyceerhl) [inline src/client/datascience/notebook/outputs/rendererCommunication.ts:36]:
```suggestion
                    this.plotSaveHandler.savePlot(editor, message.outputId, message.mimeType).catch(noop);
```

Comment (joyceerhl) [reply src/client/datascience/notebook/outputs/rendererCommunication.ts:36]: Hm, maybe I'm missing something. Where does the corresponding `postMessage` happen? I'm not seeing it in this PR. Though I assume it comes from the renderers, so maybe that change happens in https://github.com/microsoft/vscode-notebook-renderers?

Comment (joyceerhl) [reply src/client/datascience/notebook/outputs/rendererCommunication.ts:36]: Never mind, just read your issue description 🤡 

> Corresponding change microsoft/vscode-notebook-renderers#55

File: src/client/datascience/notebook/outputs/rendererCommunication.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import { notebooks } from 'vscode';
+import { IExtensionSyncActivationService } from '../../../activation/types';
+import { UseVSCodeNotebookEditorApi } from '../../../common/constants';
+import { disposeAllDisposables } from '../../../common/helpers';
+import { IDisposable } from '../../../common/types';
+import { noop } from '../../../common/utils/misc';
+import { PlotSaveHandler } from './plotSaveHandler';
+
+const rendererId = 'jupyter-notebook-renderer';
+type RendererMessageTypes = { type: 'saveAs'; outputId: string; mimnType: string };
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/outputs/rendererCommunication.ts:14]:
```suggestion
type RendererMessageTypes = { type: 'saveAs'; outputId: string; mimeType: string };
```

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/notebook/outputs/plotSaveHandler.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    ) {}
+
+    @swallowExceptions()
+    public async savePlot(editor: NotebookEditor, outputId: string, mimeType: string) {
+        if (editor.document.isClosed) {
+            return;
+        }
+        const output = getOutputItem(editor, outputId, mimeType);
+        if (!output) {
+            throw new Error(`Nolot to save ${editor.document.uri.toString()}, id: ${outputId} for ${mimeType}`);
```
Comment (IanMatthewHuff) [inline src/client/datascience/notebook/outputs/plotSaveHandler.ts:33]: Is it intentional to both @swallowExceptions and throw Errors here? Isn't that just traceError in the end?

File: src/client/datascience/notebook/outputs/rendererCommunication.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import { inject, injectable } from 'inversify';
+import { notebooks } from 'vscode';
+import { IExtensionSyncActivationService } from '../../../activation/types';
+import { UseVSCodeNotebookEditorApi } from '../../../common/constants';
+import { disposeAllDisposables } from '../../../common/helpers';
+import { IDisposable } from '../../../common/types';
+import { noop } from '../../../common/utils/misc';
+import { PlotSaveHandler } from './plotSaveHandler';
+
+const rendererId = 'jupyter-notebook-renderer';
```
Comment (IanMatthewHuff) [inline src/client/datascience/notebook/outputs/rendererCommunication.ts:13]: This is already a constant that we have. JupyterNotebookRenderer.

**Review by IanMatthewHuff — APPROVED** [review]
Approved with a couple minor comments.

File: news/1 Enhancements/6183.md [hunk 5]
```diff
@@ -0,0 +1 @@
+Added ability to save Plots.
```
Comment (joyceerhl) [inline news/1 Enhancements/6183.md:1]:
```suggestion
Added ability to save plots in the preview native notebooks UI.
```

File: src/client/datascience/notebook/outputs/plotSaveHandler.ts (hunk 3 above)
Comment (DonJayamanne) [reply src/client/datascience/notebook/outputs/plotSaveHandler.ts:33]: Lol yes, initially I wrote throw, but I didn't want the calling code to care about the exceptions, then added swallow..
Will change to just trace, that's simpler.

**Review by joyceerhl — APPROVED** [review]

File: src/client/datascience/notebook/outputs/plotSaveHandler.ts (hunk 3 above)
Comment (DonJayamanne) [reply src/client/datascience/notebook/outputs/plotSaveHandler.ts]: Done

**Pull request merged**